### PR TITLE
Refine pubsub lifecycle change conventions

### PR DESCRIPTION
**Pull request by jessemyers**
The pubsub framework defines a strong convention for publishing messages in response to CRUD
lifecycle changes. Specifically:

 -  Message media types follow a deterministic (and reversible) pattern based on the resource type
    and lifecycle change value
 -  Message schemas can be derived predictably from the media type (and lifecycle change)

These conventions were previously implemented opaquely and non-extensibly. This PR:

 -  Refactors conventions into a clearer structure
 -  Revises `LifecycleChange` as non-enumerated graph component (supporting insertion)
 -  Writes basic tests for the conventions

### CODE REVIEW

File: microcosm_pubsub/conventions.py [hunk 1]
```diff
@@ -1,148 +0,0 @@
...
-from inspect import isclass
-
-from inflection import underscore
-from marshmallow import fields
-
-from microcosm_pubsub.codecs import PubSubMessageSchema
-
-
-@unique
-class LifecycleChange(Enum):
```
Comment (jessemyers) [inline microcosm_pubsub/conventions.py]: Moved to `conventions.lifecycle`

File: microcosm_pubsub/conventions.py [hunk 2]
```diff
@@ -1,148 +0,0 @@
...
-
-    @classmethod
-    def matches(cls, media_type):
-        return any(
-            item.value in media_type.split(".")
-            for item in cls
-        )
-
-
-def name_for(obj):
```
Comment (jessemyers) [inline microcosm_pubsub/conventions.py]: Moved to `conventions.naming`

File: microcosm_pubsub/conventions.py [hunk 3]
```diff
@@ -1,148 +0,0 @@
...
-        # differentiate messages sent over a pubsub channel from other messages
-        "pubsub" if public else "pubsub._",
-        # qualify the message with the kind of lifecycle change
-        lifecycle_change.value,
-        # specify the resource name
-        name_for(resource),
-    )
-
-
-class URIMessageSchema(PubSubMessageSchema):
```
Comment (jessemyers) [inline microcosm_pubsub/conventions.py]: Moved to `conventions.messages`

File: microcosm_pubsub/conventions/__init__.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+"""
+from functools import partial
+
+from microcosm_pubsub.conventions.lifecycle import LifecycleChange
+from microcosm_pubsub.conventions.messages import IdentityMessageSchema, URIMessageSchema  # noqa: F401
+from microcosm_pubsub.conventions.naming import make_media_type, name_for  # noqa: F401
+
+
+def media_type(lifecycle_change):
```
Comment (jessemyers) [inline microcosm_pubsub/conventions/__init__.py]: This partial function can be used to generate new `verbed` wrappers. I intend to do away with `stopped` and move it to the one place our code base we use it as a custom lifecycle change.

File: microcosm_pubsub/conventions/lifecycle.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+"""
+Enumeration of common lifecycle changes.
+
+"""
+
+
+class LifecycleChange(set):
```
Comment (jessemyers) [inline microcosm_pubsub/conventions/lifecycle.py]: Now a *set*, not an *Enum*

File: microcosm_pubsub/conventions/naming.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    if organization is None:
+        organization = "globality"
+
+    return "application/vnd.{}.{}.{}.{}".format(
+        # use a vendor specific media type
+        organization,
+        # differentiate messages sent over a pubsub channel from other messages
+        "pubsub" if public else "pubsub._",
+        # qualify the message with the kind of lifecycle change
+        str(lifecycle_change),
```
Comment (jessemyers) [inline microcosm_pubsub/conventions/naming.py]: We use a string here now, not a `enum.value`

File: microcosm_pubsub/tests/conventions/test_conventions.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+"""
+from hamcrest import assert_that, equal_to, instance_of, is_
+
+from microcosm.api import create_object_graph
+from microcosm_pubsub.conventions import created, deleted, media_type
+from microcosm_pubsub.conventions.messages import IdentityMessageSchema, URIMessageSchema
+
+
+def borked(resource, **kwargs):
```
Comment (jessemyers) [inline microcosm_pubsub/tests/conventions/test_conventions.py]: Custom `verbed` wrapper

File: microcosm_pubsub/tests/conventions/test_conventions.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+        assert_that(
+            self.graph.pubsub_message_schema_registry.find(
+                "application/vnd.globality.pubsub._.deleted.foo.bar",
+            ).schema,
+            is_(instance_of(IdentityMessageSchema)),
+        )
+
+    def test_custom_lifecycle_change(self):
+        self.graph.pubsub_lifecycle_change.add("borked")
```
Comment (jessemyers) [inline microcosm_pubsub/tests/conventions/test_conventions.py]: Register `borked`

File: microcosm_pubsub/conventions/lifecycle.py (hunk 5 above)
Comment (GabeLambda) [reply microcosm_pubsub/conventions/lifecycle.py]: What's the main benefit here?

Comment (GabeLambda) [reply microcosm_pubsub/conventions/lifecycle.py]: ah, so `make_media_type` doesn't need to be tied here.

File: microcosm_pubsub/tests/conventions/test_conventions.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        )
+
+        assert_that(
+            self.graph.pubsub_message_schema_registry.find(
+                "application/vnd.globality.pubsub._.created.foo.bar",
+            ).schema,
+            is_(instance_of(URIMessageSchema)),
+        )
+
+    def test_delerted(self):
```
Comment (GabeLambda) [inline microcosm_pubsub/tests/conventions/test_conventions.py]: typo

File: microcosm_pubsub/conventions/lifecycle.py (hunk 5 above)
Comment (jessemyers) [reply microcosm_pubsub/conventions/lifecycle.py]: And it's extensible.

**Pull request closed without merging**